### PR TITLE
Fix Array `beforeEncode`/`afterDecode` for `Content`

### PR DESCRIPTION
**Pull request by 0xTim**
This fixes a bug where `beforeEncode` or `afterDecode` wouldn't be called on an array of `Content`.

However to make this work for value types, it requires replacing each element in the array which could be a significant slow down. Thoughts?

### CODE REVIEW

File: Sources/Vapor/Content/Content.swift [hunk 1]
```diff
@@ -123,6 +123,24 @@ extension Array: Content, ResponseEncodable, RequestDecodable where Element: Con
...
+            self.insert(element, at: index)
+        }
+    }
+
+    public mutating func afterDecode() throws {
+        for index in 0..<self.count {
+            var element = self[index]
+            try element.afterDecode()
+            self.remove(at: index)
+            self.insert(element, at: index)
```
Comment (madsodgaard) [inline Sources/Vapor/Content/Content.swift:141]: Can't this be simplified to `try self[index].afterDecode()`? The tests still pass.

Same for encoding

Comment (0xTim) [reply Sources/Vapor/Content/Content.swift]: Good shout - so the question becomes is it worth iterating through every array when most people (I'm assuming) don't use this

**Review by calebkleveter — APPROVED** [review]

**Review by siemensikkema — APPROVED** [review]
@0xTim The performance problem you mentioned seems to be resolved now?

**Comment by 0xTim** [comment]
@siemensikkema The worst of it correct. However it still introduces a penalty as we'd not loop through every element in an array. For large arrays that could be quite a hit. Ideally we'd gate this behind a flag in my mind

**Comment by siemensikkema** [comment]
@0xTim I guess the concern is mostly for cases where the hooks are not actually implemented and it's just iterating for no good reason. Would the compiler be able to optimize it out?
If not, without having a way to know if `Element` has an explicit implementation for these hooks we'd need some extra (static) variable on `Content` where we can control this behavior, or do you have something better in mind?
To complicate matters even more, one could say we should give `Dictionary<String, Content>` the same treatment if we're doing this.
Alternatively we leave it up the to developer to add the relevant extension for Arrays or Dictionaries where they need this behavior.

**Comment by 0xTim** [comment]
Yeah I've tried
```swift
extension Array where Element == SomeData {
    mutating func beforeEncode() throws {
        for index in 0..<self.count {
            try self[index].beforeEncode()
        }
    }
}

extension Array where Element: Content {
    mutating func beforeEncode() throws {
        for index in 0..<self.count {
            try self[index].beforeEncode()
        }
    }
}
```

Both don't get triggered

**Comment by 0xTim** [comment]
Closing as I don't think that the tradeoff is worth it

**Pull request closed without merging**